base/compiler/typelimits.jl

```diff
@@ -317,6 +317,27 @@ function tmerge(@nospecialize(typea), @nospecialize(typeb))
         end
         return Bool
     end
+    if (isa(typea, PartialStruct) || isa(typea, Const)) &&
+       (isa(typeb, PartialStruct) || isa(typeb, Const)) &&
+        widenconst(typea) === widenconst(typeb)
```

Comment:
It's unclear to me that this tmerge relaxation is desirable, since it could be much slower. But it should at least have unit tests for all of the new cases being added here.